### PR TITLE
update `Number` class to handle integer values

### LINKED ISSUE #8153: [CT-2848] [Bug] `dbt show` adding decimal places to non-decimal values
### Is this a regression in a recent version of dbt-core?

- [X] I believe this is a regression in dbt-core functionality
- [X] I have searched the existing issues, and I could not find an existing issue for this regression

### Current Behavior

`dbt show` is adding decimal places to numeric/integer values incorrectly. This is causing problems for users in the dbt Cloud IDE, as the IDE uses that command to render the preview tab. 

### Expected/Previous Behavior

Previously (we think!) this was not an issue, and decimal places were properly handled by dbt show

### Steps To Reproduce

1. use dbt-snowflake>=1.5.0
2. create a model in your project:

```sql
# in my_model.sql
select
  cast(1 as numeric) as _numeric,
  cast(1 as integer) as _integer,
  cast(1 as decimal) as _decimal
```
3. run `dbt show -s my_model --output JSON`
4. get this!
```zsh
❯ dbt show -s my_model --output json
20:33:15  Running with dbt=1.5.3
20:33:16  Registered adapter: snowflake=1.5.2
20:33:16  Unable to do partial parsing because of a version mismatch
20:33:17  Found 3 models, 4 tests, 0 snapshots, 0 analyses, 436 macros, 0 operations, 0 seed files, 0 sources, 0 exposures, 0 metrics, 0 groups
20:33:17  
20:33:18  Concurrency: 8 threads (target='dev')
20:33:18  
20:33:19  {
  "node": "my_model",
  "show": [
    {
      "_NUMERIC": 1.0,
      "_INTEGER": 1.0,
      "_DECIMAL": 1.0
    }
  ]
}
```

### Relevant log output

_No response_

### Environment

```markdown
- OS: mac
- Python: Python 3.9.16
- dbt (working version): unknown
- dbt (regression version): 1.5.3
```

### Which database adapter are you using with dbt?

snowflake

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by dave-connors-3**
resolves #8153

### Problem

Agate tables, which power `dbt show`, and therefore the dbt cloud IDE, show integer values as decimals, leading to user confusion when columns seem to have decimals when they don't! this overrides that behavior in the dbt-core customization of the agate `Number` class

### Solution

Ensure that the `Number.cast()` and `Number.jsonify()` methods properly handle integer values. 

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me  
- [x] I have run this code in development and it appears to resolve the stated issue  
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [ ] This PR has no interface changes (e.g. macros, cli, logs, json artifacts, config files, adapter interface, etc) or this PR has already received feedback and approval from Product or DX

### CODE REVIEW

File: tests/functional/show/fixtures.py [hunk 1]
```diff
@@ -75,10 +83,10 @@
 
 seeds__sample_seed = """sample_num,sample_bool
 1,true
-2,false
+2.0,false
```
Comment (dave-connors-3) [inline tests/functional/show/fixtures.py:86]: shouldn't impact anything, but this is not a necessary change, can roll it back

**Comment by jtcohen6** [comment]
Thanks for putting up the PR @dave-connors-3!

The CI failure reminded me that my recommendation was incomplete.

We need to choose whether we want to cast to `int` sooner or later:
- Later: just during JSON-ification
- Sooner: as in this approach, but taken one step further to a custom `Integer` type

I left a comment on the issue with two potential approaches: https://github.com/dbt-labs/dbt-core/issues/8153#issuecomment-1666889641

**Comment by emmyoop** [comment]
@dave-connors-3 I have capacity this sprint to help get this over this finish line!  I agree with @jtcohen6's recommendation to add a new Integer data type, as laid out in his comment.  It's the right solution, even with the risk of being a breaking change.  Let me know if you'd like some help or even if you don't have capacity right now and I can help get this in.

**Comment by dave-connors-3** [comment]
thanks @emmyoop!! just pushed jerco's updates -- will test locally shortly, but would love your eyes on it whenever you have a chance!

**Comment by dave-connors-3** [comment]
@emmyoop -- just got the test_agate_helper.py` tests to pass locally! I am wondering if the other tests should have been updated to assert that they were instances of the custom Integer class as well, but updating them caused local failures. Let me know what you think!

File: tests/unit/test_agate_helper.py [hunk 2]
```diff
@@ -121,37 +121,37 @@ def test_datetime_formats(self):
...
-        t1 = agate.Table([(1, "a", None), (2, "b", None)], ("a", "b", "c"))
-        t2 = agate.Table([(3, "c", None), (4, "d", None)], ("a", "b", "c"))
+        t1 = agate_helper.table_from_rows([(1, "a", None), (2, "b", None)], ("a", "b", "c"))
+        t2 = agate_helper.table_from_rows([(3, "c", None), (4, "d", None)], ("a", "b", "c"))
         result = agate_helper.merge_tables([t1, t2])
         self.assertEqual(result.column_names, ("a", "b", "c"))
-        assert isinstance(result.column_types[0], agate.data_types.Number)
+        assert isinstance(result.column_types[0], agate_helper.Integer)
         assert isinstance(result.column_types[1], agate.data_types.Text)
         assert isinstance(result.column_types[2], agate.data_types.Number)
```
Comment (dave-connors-3) [inline tests/unit/test_agate_helper.py:130]: @emmyoop this one wanted to stay a `Number` -- guessing it has something to do with the fact that the column is null, but wasn't positive!

**Comment by dave-connors-3** [comment]
i also seem to be cursed by the codecoverage bots

File: .changes/unreleased/Fixes-20230803-093502.yaml [hunk 3]
```diff
@@ -0,0 +1,6 @@
+kind: Fixes
+body: update agage.Number to handle ints
```
Comment (emmyoop) [inline .changes/unreleased/Fixes-20230803-093502.yaml:2]:
```suggestion
body: Add explicit support for integers for the show command
```

**Review by emmyoop — APPROVED** [review]
This looks great!  Thanks so much got getting this in @dave-connors-3!

**Pull request merged**